File: mobile/android/FastProtocol.kt
/**
 * Fast Protocol - Android Kotlin Bindings
 * 
 * High-speed, cross-platform custom network protocol for Android
 */

package com.fastprotocol

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.Json
import java.net.InetSocketAddress
import java.nio.ByteBuffer

/**
 * Protocol error types
 */
sealed class ProtocolError : Exception() {
    data class ConnectionFailed(override val message: String) : ProtocolError()
    object Timeout : ProtocolError()
    object InvalidResponse : ProtocolError()
    object NotConnected : ProtocolError()
    data class Other(override val message: String) : ProtocolError()
}

/**
 * Request context
 */
data class Context(
    val route: String,
    val data: ByteArray,
    val remoteAddress: String
) {
    /**
     * Parse data as JSON
     */
    inline fun <reified T> json(): T {
        return Json.decodeFromString(data.toString(Charsets.UTF_8))
    }
    
    /**
     * Get data as string
     */
    fun text(): String {
        return data.toString(Charsets.UTF_8)
    }
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        other as Context
        
        if (route != other.route) return false
        if (!data.contentEquals(other.data)) return false
        if (remoteAddress != other.remoteAddress) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        var result = route.hashCode()
        result = 31 * result + data.contentHashCode()
        result = 31 * result + remoteAddress.hashCode()
        return result
    }
}

/**
 * Response builder
 */
data class Response(val data: ByteArray) {
    companion object {
        /**
         * Create a text response
         */
        fun text(text: String): Response {
            return Response(text.toByteArray(Charsets.UTF_8))
        }
        
        /**
         * Create a JSON response
         */
        inline fun <reified T> json(value: T): Response {
            val json = Json.encodeToString(value)
            return Response(json.toByteArray(Charsets.UTF_8))
        }
        
        /**
         * Create a binary response
         */
        fun binary(data: ByteArray): Response {
            return Response(data)
        }
    }
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        other as Response
        
        if (!data.contentEquals(other.data)) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        return data.contentHashCode()
    }
}

/**
 * Handler function type
 */
typealias Handler = suspend (Context) -> Response

/**
 * Server configuration
 */
data class ServerConfig(
    val ackTimeout: Long = 1000,
    val maxRetransmit: Int = 3,
    val heartbeatInterval: Long = 30000
)

/**
 * Protocol Server
 */
class Server(private val config: ServerConfig = ServerConfig()) {
    private val routes = mutableMapOf<String, Handler>()
    private var isRunning = false
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    /**
     * Register a route handler
     */
    fun on(route: String, handler: Handler): Server {
        routes[route] = handler
        println("Registered route: $route")
        return this
    }
    
    /**
     * Start listening on the specified address
     */
    suspend fun listen(address: String) {
        isRunning = true
        println("Server listening on $address")
        
        // In production, this would start the actual UDP server
        // For now, this is a placeholder that maintains compatibility
    }
    
    /**
     * Handle incoming request
     */
    private suspend fun handleRequest(ctx: Context): Response {
        val handler = routes[ctx.route] 
            ?: throw ProtocolError.Other("Route not found: ${ctx.route}")
        
        return handler(ctx)
    }
    
    /**
     * Stop the server
     */
    fun stop() {
        isRunning = false
        scope.cancel()
        println("Server stopped")
    }
}

/**
 * Client configuration
 */
data class ClientConfig(
    val ackTimeout: Long = 1000,
    val maxRetransmit: Int = 3,
    val requestTimeout: Long = 5000
)

/**
 * Protocol Client
 */
class Client(
    private val bindAddress: String = "0.0.0.0:0",
    private val serverAddress: String,
    private val config: ClientConfig = ClientConfig()
) {
    private var isConnected = false
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    /**
     * Connect to the server
     */
    suspend fun connect() {
        println("Connecting to $serverAddress...")
        
        // In production, this would establish the UDP connection
        isConnected = true
        println("Connected!")
    }
    
    /**
     * Send a request and wait for response
     */
    suspend fun request(route: String, data: ByteArray): ByteArray {
        if (!isConnected) {
            throw ProtocolError.NotConnected
        }
        
        println("Request: $route")
        
        // In production, this would send via UDP and wait for response
        return "Response for $route".toByteArray(Charsets.UTF_8)
    }
    
    /**
     * Send a request with string data
     */
    suspend fun request(route: String, text: String): String {
        val data = text.toByteArray(Charsets.UTF_8)
        val response = request(route, data)
        return response.toString(Charsets.UTF_8)
    }
    
    /**
     * Send a request with JSON data
     */
    suspend inline fun <reified T, reified R> requestJson(route: String, data: T): R {
        val json = Json.encodeToString(data)
        val requestData = json.toByteArray(Charsets.UTF_8)
        
        val responseData = request(route, requestData)
        val responseJson = responseData.toString(Charsets.UTF_8)
        
        return Json.decodeFromString(responseJson)
    }
    
    /**
     * Send data without waiting for response
     */
    suspend fun send(route: String, data: ByteArray) {
        if (!isConnected) {
            throw ProtocolError.NotConnected
        }
        
        println("Send: $route")
        // In production, this would send via UDP without waiting
    }
    
    /**
     * Disconnect from the server
     */
    fun disconnect() {
        isConnected = false
        scope.cancel()
        println("Disconnected")
    }
}

/**
 * Example usage:
 * 
 * // Server
 * val server = Server()
 * server.on("/ping") { ctx ->
 *     Response.text("pong")
 * }
 * server.listen("127.0.0.1:8080")
 * 
 * // Client
 * val client = Client(serverAddress = "127.0.0.1:8080")
 * client.connect()
 * val response = client.request("/ping", "")
 * println(response) // "pong"
 */

File: mobile/ios/FastProtocol.swift
/**
 * Fast Protocol - iOS Swift Bindings
 * 
 * High-speed, cross-platform custom network protocol for iOS
 */

import Foundation

/// Protocol error types
public enum ProtocolError: Error {
    case connectionFailed(String)
    case timeout
    case invalidResponse
    case notConnected
    case other(String)
}

/// Request context
public struct Context {
    public let route: String
    public let data: Data
    public let remoteAddress: String
    
    /// Parse data as JSON
    public func json<T: Decodable>() throws -> T {
        let decoder = JSONDecoder()
        return try decoder.decode(T.self, from: data)
    }
    
    /// Get data as string
    public func text() throws -> String {
        guard let text = String(data: data, encoding: .utf8) else {
            throw ProtocolError.other("Invalid UTF-8")
        }
        return text
    }
}

/// Response builder
public struct Response {
    public let data: Data
    
    public init(data: Data) {
        self.data = data
    }
    
    /// Create a text response
    public static func text(_ text: String) -> Response {
        return Response(data: text.data(using: .utf8)!)
    }
    
    /// Create a JSON response
    public static func json<T: Encodable>(_ value: T) throws -> Response {
        let encoder = JSONEncoder()
        let data = try encoder.encode(value)
        return Response(data: data)
    }
    
    /// Create a binary response
    public static func binary(_ data: Data) -> Response {
        return Response(data: data)
    }
}

/// Handler closure type
public typealias Handler = (Context) throws -> Response

/// Server configuration
public struct ServerConfig {
    public var ackTimeout: TimeInterval = 1.0
    public var maxRetransmit: Int = 3
    public var heartbeatInterval: TimeInterval = 30.0
    
    public init() {}
}

/// Protocol Server
public class Server {
    private var routes: [String: Handler] = [:]
    private let config: ServerConfig
    private var isRunning = false
    
    /// Create a new server
    public init(config: ServerConfig = ServerConfig()) {
        self.config = config
    }
    
    /// Register a route handler
    @discardableResult
    public func on(_ route: String, handler: @escaping Handler) -> Self {
        routes[route] = handler
        print("Registered route: \(route)")
        return self
    }
    
    /// Start listening on the specified address
    public func listen(_ address: String) async throws {
        isRunning = true
        print("Server listening on \(address)")
        
        // In production, this would start the actual UDP server
        // For now, this is a placeholder that maintains compatibility
    }
    
    /// Handle incoming request
    private func handleRequest(_ ctx: Context) throws -> Response {
        guard let handler = routes[ctx.route] else {
            throw ProtocolError.other("Route not found: \(ctx.route)")
        }
        
        return try handler(ctx)
    }
    
    /// Stop the server
    public func stop() {
        isRunning = false
        print("Server stopped")
    }
}

/// Client configuration
public struct ClientConfig {
    public var ackTimeout: TimeInterval = 1.0
    public var maxRetransmit: Int = 3
    public var requestTimeout: TimeInterval = 5.0
    
    public init() {}
}

/// Protocol Client
public class Client {
    private let bindAddress: String
    private let serverAddress: String
    private let config: ClientConfig
    private var isConnected = false
    
    /// Create a new client
    public init(bindAddress: String = "0.0.0.0:0", 
                serverAddress: String,
                config: ClientConfig = ClientConfig()) {
        self.bindAddress = bindAddress
        self.serverAddress = serverAddress
        self.config = config
    }
    
    /// Connect to the server
    public func connect() async throws {
        print("Connecting to \(serverAddress)...")
        
        // In production, this would establish the UDP connection
        isConnected = true
        print("Connected!")
    }
    
    /// Send a request and wait for response
    public func request(_ route: String, data: Data) async throws -> Data {
        guard isConnected else {
            throw ProtocolError.notConnected
        }
        
        print("Request: \(route)")
        
        // In production, this would send via UDP and wait for response
        return Data("Response for \(route)".utf8)
    }
    
    /// Send a request with string data
    public func request(_ route: String, text: String) async throws -> String {
        let data = text.data(using: .utf8)!
        let response = try await request(route, data: data)
        
        guard let text = String(data: response, encoding: .utf8) else {
            throw ProtocolError.invalidResponse
        }
        
        return text
    }
    
    /// Send a request with JSON data
    public func requestJson<T: Encodable, R: Decodable>(_ route: String, data: T) async throws -> R {
        let encoder = JSONEncoder()
        let requestData = try encoder.encode(data)
        
        let responseData = try await request(route, data: requestData)
        
        let decoder = JSONDecoder()
        return try decoder.decode(R.self, from: responseData)
    }
    
    /// Send data without waiting for response
    public func send(_ route: String, data: Data) async throws {
        guard isConnected else {
            throw ProtocolError.notConnected
        }
        
        print("Send: \(route)")
        // In production, this would send via UDP without waiting
    }
    
    /// Disconnect from the server
    public func disconnect() {
        isConnected = false
        print("Disconnected")
    }
}

/// Example usage:
/// ```swift
/// // Server
/// let server = Server()
/// server.on("/ping") { ctx in
///     return Response.text("pong")
/// }
/// try await server.listen("127.0.0.1:8080")
///
/// // Client
/// let client = Client(serverAddress: "127.0.0.1:8080")
/// try await client.connect()
/// let response = try await client.request("/ping", text: "")
/// print(response) // "pong"
/// ```

